src/slurm_plugin/common.py

```diff
@@ -8,5 +8,183 @@
 # OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
 # limitations under the License.
+import collections
+import logging
+import re
+import subprocess
+
+import boto3
+from botocore.exceptions import ClientError
+
+from common.schedulers.slurm_commands import update_nodes
+from common.utils import grouper
 
 CONFIG_FILE_DIR = "/opt/parallelcluster/configs/slurm"
+
+log = logging.getLogger(__name__)
+
+
+def terminate_associated_instances(slurm_nodes, region, cluster_name, boto3_config, batch_size):
+    """Terminate instances associated with given nodes in batches."""
+    instance_ids_to_nodename = _get_instance_ids_to_nodename(slurm_nodes, region, cluster_name, boto3_config)
+    log.debug("instance_ids_to_nodename = %s", instance_ids_to_nodename)
+    if instance_ids_to_nodename:
+        delete_instances(
+            list(instance_ids_to_nodename.keys()), region, boto3_config, batch_size,
+        )
+
+
+def _get_instance_ids_to_nodename(slurm_nodes, region, cluster_name, boto3_config):
+    """Retrieve dict that maps from instance ids to slurm nodenames."""
+    node_ip_to_name = {node.nodeaddr: node.name for node in slurm_nodes}
+    ec2_client = boto3.client("ec2", region_name=region, config=boto3_config)
+    paginator = ec2_client.get_paginator("describe_instances")
+    response_iterator = paginator.paginate(
+        Filters=[
+            {"Name": "private-ip-address", "Values": list(node_ip_to_name.keys())},
+            {"Name": "tag:ClusterName", "Values": [cluster_name]},
+        ],
+    )
+    filtered_iterator = response_iterator.search("Reservations[].Instances[]")
+    return {
+        instance_info["InstanceId"]: node_ip_to_name[instance_info["PrivateIpAddress"]]
+        for instance_info in filtered_iterator
+    }
+
+
+def delete_instances(instance_ids_to_terminate, region, boto3_config, batch_size):
+    """Terminate corresponding EC2 instances."""
+    ec2_client = boto3.client("ec2", region_name=region, config=boto3_config)
+    log.info("Terminating instances %s", instance_ids_to_terminate)
+    for instances in grouper(instance_ids_to_terminate, batch_size):
+        try:
+            # Boto3 clients retries on connection errors only
+            ec2_client.terminate_instances(InstanceIds=list(instances),)
+        except ClientError as e:
+            log.error("Failed when terminating instances %s with error %s", instances, e)
+
+
+class InstanceLauncher:
+    """
+    InstanceLauncher class.
+
+    Class to manage instance launching action.
+    Should generally be used when launching instances for slurm integration.
+    """
+
+    def __init__(self, node_list, region, cluster_name, boto3_config, max_batch_size, update_node_address):
+        """Initialize InstanceLauncher with required attributes."""
+        self.node_list = node_list
```

Comment:
let's stick to python naming convention by adding a leading underscore to class variables that are not part of the public class interface